R/convert-gene-ids.R

```diff
@@ -0,0 +1,125 @@
+#' Convert Ensembl gene ids to gene symbols based on an ScPCA SingleCellExperiment object
+#'
+#' The SingleCellExperiment objects produced as part of ScPCA are indexed by
+#' Ensembl gene ids, as those are more stable than gene symbols. However,
+#' for many applications gene symbols are useful. This function provides a
+#' simple and consistent conversion of Ensembl gene ids to gene symbols based on
+#' the `gene_symbol` column that is present in the row data of ScPCA
+#' SingleCellExperiment objects.
+#'
+#' For this function, the SingleCellExperiment object must contain a `gene_ids`
+#' column containing Ensembl gene ids and a `gene_symbol` column containing gene
+#' symbols. If any gene ids are not found or if the gene symbol is not defined,
+#' the input gene id is returned, unless the `leave_na` is set to `TRUE`.
+#'
+#'
+#' @param ensembl_ids A character vector of Ensembl gene ids to translate to
+#'  gene symbols.
+#' @param sce A SingleCellExperiment object containing gene ids and gene symbols
+#'  to use for translation.
+#' @param leave_na logical indicating whether to leave NA values in the output.
+#'  Default is `FALSE`
+#'
+#' @return A vector of gene symbols corresponding to the input Ensembl ids.
+#' @export
+#'
+#' @import SingleCellExperiment
+#'
+#' @examples
+#' \dontrun{
+#' # convert a set of Ensembl ids to gene symbols
+#' # using a SingleCellExperiment reference
+#' ensembl_ids <- c("ENSG00000141510", "ENSG00000134323")
+#' gene_symbols <- ensembl_to_symbol(ensembl_ids, sce)
+#' gene_symbols
+#' ### [1] "TP53" "MYCN"
+#' }
+ensembl_to_symbol <- function(ensembl_ids, sce, leave_na = FALSE) {
+  stopifnot(
+    "`sce` must be a SingleCellExperiment object ." = is(sce, "SingleCellExperiment"),
+    "`ensembl_ids` must be a character vector." = is.character(ensembl_ids),
+    "`sce` must contain both a `gene_ids` and `gene_symbol` column in the row data." =
+      all(c("gene_ids", "gene_symbol") %in% names(rowData(sce))),
+    "`leave_na` must be TRUE or FALSE." = is.logical(leave_na)
+  )
+
+  id_match <- match(ensembl_ids, rowData(sce)$gene_ids)
+  gene_symbols <- rowData(sce)[id_match, "gene_symbol"]
+
+  missing_symbols <- is.na(gene_symbols)
+  if (!leave_na && any(missing_symbols)) {
+    warning("Not all `ensembl_ids` are present in sce `gene_ids`, using input ids for missing values.")
```

Comment:
The input id may not actually be an Ensembl id...

Reply:
Only if we change the argument name, which I don't want to do because it tells people the kind of ids that are expected.

Reply:
I won't die on this hill, but this confuses me even more...
- we tell them ensembl ids are "expected" via argument/variable names
- they don't have to provide ensembl ids since the code is flexible enough
- if they don't provide ensembl ids, the error will tell them there's a "problem" with their ensembl ids. If I were in this camp, the error message would confuse me!

Reply:
We are telling them there is a problem with the values they gave to the `ensembl_ids` argument.  It isn't saying that there was a necessarily problem with the ids, but that there may not be a gene symbol for the id.

Reply:
Which is actually a problem with the error message, now that I read it. How about: 
```suggestion
    warning("Not all `ensembl_ids` values have corresponding gene symbols, using input ids for missing values.")
```

Reply:
Ah ok ok ok - since the argument is called `ensembl_ids` to begin with (that's not _only_ an internal variable name is the thing I glanced over!), they know what they've done. ✅